faq.md

```diff
@@ -499,6 +499,81 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
+
+* Event-driven programming: User interactions are represented by event objects that are
+dispatched to callbacks registered with event handlers. Screen updates are triggered
+by a similar callback mechanism. The [Listenable](https://docs.flutter.io/flutter/foundation/Listenable-class.html)
+class, which is used as the basis of the animation system, formalises a
+subscription model for events with multiple listeners.
+
+* Class-based object-orientated programming: Most of the APIs of the framework are built using classes
+with inheritance. We use an approach whereby we define very high-level APIs in our base classes, then
+specialise them iteratively in subclasses. For example, our render objects have a base class
+([RenderObject](https://docs.flutter.io/flutter/rendering/RenderObject-class.html))
+that is agnostic regarding the coordinate system, and then we have a subclass
+([RenderBox](https://docs.flutter.io/flutter/rendering/RenderBox-class.html))
+that introduces the opinion that the geometry should be based on the Cartesian coordinate system (x/width and y/height).
+
+* Prototype-based object-orientated programming: The
+[ScrollPhysics](https://docs.flutter.io/flutter/widgets/ScrollPhysics-class.html) class
+chains instances to compose the physics that apply to scrolling dynamically at runtime. This
+lets the system compose, for example, paging physics with platform-specific physics, without
+the platform having to be selected at compile time.
+
+* Imperative programming: Straight forward imperative programming, usually paired with state encapsulated
+within an object, is used where it provides the most intuitive solution. For example, tests are
+written in an imperative style, first describing the situation under test, then listing the invariants
+that the test must match, then advancing the clock or inserting events as necessary for the test.
+
+* Reactive programming: The widget and element trees are sometimes described as reactive, since new inputs 
+provided in a widget's constructor are immediately propagated as changes to lower-level widgets by
+the widget's build method, and changes made in the lower widgets (e.g. in response to user input) propagate
+back up the tree via event handlers. Aspects of both functional-reactive and imperative-reactive are
+present in the framework.
+
+* Declarative programming: The build functions of widgets are typically a single expression with multiple levels
+of nested constructors, very much a declarative approach to building UI. This can also be combined with
+the imperative style to build UIs that would be harder to describe in a pure-declarative approach.
+
+* Generic programming: Types can be used to help catch programming errors early. The Flutter framework
+uses generic programming to help in this regard. For example, the
+[State](https://docs.flutter.io/flutter/widgets/State-class.html) class is parameterized in terms
+of the type of its associated widget, so that it 
```

Comment:
oops! Added: ...so that the Dart analyzer can catch mismatches of states and widgets. Similarly, the [GlobalKey](https://docs.flutter.io/flutter/widgets/GlobalKey-class.html) class takes a type parameter so that it can access a remote widget's state in a type-safe manner (with runtime checking), the [Route](https://docs.flutter.io/flutter/widgets/Route-class.html) interface is parameterized with the type that it is expected to use when [popped](https://docs.flutter.io/flutter/widgets/Navigator/pop.html), and collections such as [List](https://docs.flutter.io/flutter/dart-core/List-class.html)s, [Map](https://docs.flutter.io/flutter/dart-core/Map-class.html)s, and [Set](https://docs.flutter.io/flutter/dart-core/Set-class.html)s are all parameterized so that mismatched elements can be caught early either by the analyzer or at runtime during debugging.